### PR TITLE
Check database version mismatch

### PR DESCRIPTION
**Pull request by MarinPostma**
Checks if the versions of the database and the engine are compatible.

The database and the engine are compatible if they share the same major and minor version.

The engine will refuse to start if there is a mismatch.

@bidoubiwa do we need to document this?

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -1,3 +1,7 @@
+## v0.12.0
```
Comment (Kerollmops) [inline CHANGELOG.md:1]:
```suggestion
## v0.12.0 (unreleased)
```

File: meilisearch-core/src/database.rs [hunk 2]
```diff
@@ -162,11 +163,62 @@ fn update_awaiter(
...
+/// Ensures Meilisearch version is compatible with the database, returns an error versions mismatch.
+/// If create is set to true, a VERSION file is created with the current version.
+fn version_guard(path: &Path, create: bool) -> MResult<()> {
+    let current_version_major = env!("CARGO_PKG_VERSION_MAJOR");
+    let current_version_minor = env!("CARGO_PKG_VERSION_MINOR");
+    let current_version_patch = env!("CARGO_PKG_VERSION_PATCH");
+    let version_path = path.join("VERSION");
+
+    match File::open(&version_path) {
+        Ok(mut file) => {
```
Comment (Kerollmops) [inline meilisearch-core/src/database.rs:175]: Ho! No, could you write it inside of LMDB using heed, please?

File: meilisearch-core/src/database.rs [hunk 3]
```diff
@@ -162,11 +163,62 @@ fn update_awaiter(
...
 impl Database {
     pub fn open_or_create(path: impl AsRef<Path>, options: DatabaseOptions) -> MResult<Database> {
         let main_path = path.as_ref().join("main");
         let update_path = path.as_ref().join("update");
 
+        //create db directory
+        fs::create_dir_all(&path)?;
+
+        // create file only if main db wasn't created before (first run)
+        version_guard(path.as_ref(), !main_path.exists() && !update_path.exists())?;
```
Comment (Kerollmops) [inline meilisearch-core/src/database.rs:220]: Would be better to check if the `main_path` exists and write/check the version using heed.

File: meilisearch-core/src/error.rs [hunk 4]
```diff
@@ -15,22 +15,23 @@ pub type MResult<T> = Result<T, Error>;
...
-    Heed(heed::Error),
-    Fst(fst::Error),
+    SchemaMissing,
     SerdeJson(SerdeJsonError),
-    Bincode(bincode::Error),
     Serializer(SerializerError),
-    Deserializer(DeserializerError),
-    FilterParseError(PestError<Rule>),
-    FacetError(FacetError),
+    VersionMismatch(String),
```
Comment (Kerollmops) [inline meilisearch-core/src/error.rs:33]:
```suggestion
    VersionMismatch { expected: (u32, u32), found: (u32, u32) },
```

File: meilisearch-http/src/data.rs [hunk 5]
```diff
@@ -66,7 +67,13 @@ impl Data {
 
         let http_payload_size_limit = opt.http_payload_size_limit;
 
-        let db = Arc::new(Database::open_or_create(opt.db_path, db_opt).unwrap());
+        let db = match  Database::open_or_create(opt.db_path, db_opt) {
+            Ok(db) => Arc::new(db),
+            Err(e) => {
+                error!("{}", e);
+                std::process::exit(1);
```
Comment (Kerollmops) [inline meilisearch-http/src/data.rs:74]: I don't think this is the best way do do so, but let's keep this like that for now :)

File: meilisearch-core/src/error.rs [hunk 6]
```diff
@@ -138,22 +140,23 @@ impl fmt::Display for Error {
...
-            Heed(e) => write!(f, "heed error; {}", e),
-            Fst(e) => write!(f, "fst error; {}", e),
+            SchemaMissing => write!(f, "this index does not have a schema"),
             SerdeJson(e) => write!(f, "serde json error; {}", e),
-            Bincode(e) => write!(f, "bincode error; {}", e),
             Serializer(e) => write!(f, "serializer error; {}", e),
-            Deserializer(e) => write!(f, "deserializer error; {}", e),
-            FilterParseError(e) => write!(f, "error parsing filter; {}", e),
-            FacetError(e) => write!(f, "error processing facet filter: {}", e),
+            VersionMismatch(version) => write!(f, "Cannot open database, expected Meilisearch version: {}", version),
```
Comment (Kerollmops) [inline meilisearch-core/src/error.rs:158]: Can you cmore clearer and tell the user that this is the expected **client** version

File: meilisearch-core/src/database.rs [hunk 7]
```diff
@@ -162,11 +163,62 @@ fn update_awaiter(
...
+            if version_major != current_version_major || version_minor != current_version_minor {
+                return Err(Error::VersionMismatch(format!("{}.{}.XX", version_major, version_minor)));
+            }
+        }
+        Err(error) => {
+            match error.kind() {
+                ErrorKind::NotFound => {
+                    if create {
+                        // when no version file is found, and we've beem told to create one,
+                        // create a new file wioth the current version in it.
```
Comment (Kerollmops) [inline meilisearch-core/src/database.rs:192]:
```suggestion
                        // when no version file is found, and we've been told to create one,
                        // create a new file with the current version in it.
```

File: meilisearch-core/src/database.rs [hunk 8]
```diff
@@ -162,11 +163,62 @@ fn update_awaiter(
...
+                        let mut version_file = File::create(&version_path)?;
+                        version_file.write_all(format!("{}.{}.{}",
+                                current_version_major,
+                                current_version_minor,
+                                current_version_patch).as_bytes())?;
+                    } else {
+                        // when no version file is found and we were not told to create one, this
+                        // means that the version is inferior to the one this feature was added.
```
Comment (Kerollmops) [inline meilisearch-core/src/database.rs:200]:
```suggestion
                        // means that the version is inferior to the one this feature was added in.
```

**Comment by bidoubiwa** [comment]
Hey, 

I think it is a good idea to document is. I think two places could benefit this information:
- The `error codes` and a description explainaing why MeiliSearch refuses to start
- In the `Installation` guide we could add a sub-chapter called `update`. Where we explain how to update you version of MeiliSearch.

>The database and the engine are compatible if they share the same major and minor version.
The engine will refuse to start if there is a mismatch.

Does this mean that on every update of MeiliSearch the user needs to re-index all its data?
Or is the database updatable aswell ?

**Comment by MarinPostma** [comment]
@bidoubiwa Sorry, I forgot to answer. The meili will refuse to start when the major version changed, i.e when there are breaking changes.

**Comment by MarinPostma** [comment]
bors merge

**Comment by davidk01** [comment]
Question. How do I upgrade from 0.13 to 0.14? When I try to start 0.14 it fails with an error message about incorrect database versions.

**Pull request merged**